### DISCUSSION TITLE
Strange problem with RockyLinux: Signature validation failed, looking for another key

### DISCUSSION THREAD

**Discussion by jfossheim-skyfritt** [strongswan/strongswan#2093] [Q&A] [answered]
hello everyone, 

I am struggling with a to me absurd problem with strongswan on one of our RockyLinux Laptops. We have a working setup with a Mikrotik VPN gateway, wit LetsEncrypt-ceriticates, and EAP authentication. All clients windows, mac, ios, android and Linux (Debian and Fedora), are able to connect to the VPN gateway. And the RockyLinux laptop in question are able to connect to a similar Gateway running pfsense also with letsencrypt certificates. 

I have tried the prebuilt package from EPEL-repository (5.9.10-1.el9), and I have tried to recompile the latest version from source (Linux strongSwan U5.9.13/K5.14.0-362.13.1.el9_3.x86_64), and both gives me the same error. 

This is the output from a working client connecting from a Ferdora-laptop with strongswan (5.9.11), this is running via NetworkManager and charon-nm, but we get the same result using ipsec.conf / ipsec.secrets via the CLI: 

```
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG] checking certificate status of "[CN=CERTIFICATE COMMON NAME]"
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG]   requesting ocsp status from 'http://r3.o.lencr.org' ...
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG]   ocsp response correctly signed by "C=US, O=Let's Encrypt, CN=R3"
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG]   ocsp response is valid: until Feb 16 00:03:58 2024
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG] certificate status is good
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG] checking certificate status of "C=US, O=Let's Encrypt, CN=R3"
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG]   fetching crl from 'http://x1.c.lencr.org/' ...
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG]   using trusted certificate "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG]   crl correctly signed by "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG]   crl is valid: until Mar 12 00:59:59 2024
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[CFG] certificate status is good
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[IKE] authentication of '[CN=CERTIFICATE COMMON NAME]' with RSA signature successful
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[IKE] server requested EAP_IDENTITY (id 0x00), sending 'EAP-USER@DOMAIN'
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[ENC] generating IKE_AUTH request 2 [ EAP/RES/ID ]
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[NET] sending packet: from XXX.XXX.XXX.XXX[57179] to XXX.XXX.XXX.XXX[4500] (112 bytes)
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 12[NET] received packet: from YYY.YYY.YYY.YYY[4500] to YYY.YYY.YYY.YYY[57179] (144 bytes)
```

And here is the output from the RockyLinux laptop that are failing to connect with the mikrotik Gateway: 

```
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[IKE]  received end entity cert "[CN=CERTIFICATE COMMON NAME]""
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   using certificate "[CN=CERTIFICATE COMMON NAME]"
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   using untrusted intermediate certificate "C=US, O=Let's Encrypt, CN=R3"
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   using trusted ca certificate "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG] certificate policy 2.23.140.1.2.1 for '[CN=CERTIFICATE COMMON NAME]' not allowed by trustchain, ignored
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   reached self-signed root ca with a path length of 1
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG] checking certificate status of "[CN=CERTIFICATE COMMON NAME]"
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   ocsp response is valid: until Feb 15 00:58:58 2024
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   using cached ocsp response
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG] certificate status is good
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG] checking certificate status of "C=US, O=Let's Encrypt, CN=R3"
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   using trusted certificate "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   crl correctly signed by "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   crl is valid: until Mar 12 00:59:59 2024
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG]   using cached crl
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[CFG] certificate status is good
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[IKE] signature validation failed, looking for another key
Feb 10 09:43:20 [ROCKY-LAPTOP] charon[740492]: 14[ENC] generating INFORMATIONAL request 2 [ N(AUTH_FAILED) ]
```

And here is the output when the Rocky Linux laptop successfully connects to the pfsense-gateway: 

```
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[IKE] received end entity cert "[CN=CERTIFICATE COMMON NAME]"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[IKE] received issuer cert "C=US, O=Let's Encrypt, CN=R3"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   using certificate "[CN=CERTIFICATE COMMON NAME]"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   using untrusted intermediate certificate "C=US, O=Let's Encrypt, CN=R3"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   using trusted ca certificate "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG] certificate policy 2.23.140.1.2.1 for 'CN=242.62-97-222.bkkb.no' not allowed by trustchain, ignored
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   reached self-signed root ca with a path length of 1
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG] checking certificate status of "[CN=CERTIFICATE COMMON NAME]"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   requesting ocsp status from 'http://r3.o.lencr.org' ...
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   ocsp response correctly signed by "C=US, O=Let's Encrypt, CN=R3"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   ocsp response is valid: until Feb 17 09:57:58 2024
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG] certificate status is good
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG] checking certificate status of "C=US, O=Let's Encrypt, CN=R3"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   fetching crl from 'http://x1.c.lencr.org/' ...
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   using trusted certificate "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   crl correctly signed by "C=US, O=Internet Security Research Group, CN=ISRG Root X1"
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG]   crl is valid: until Mar 12 00:59:59 2024
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[CFG] certificate status is good
Feb 10 09:58:34 [ROCKY-LAPTOP] charon[787301]: 13[IKE] authentication of '[HOSTNAME (SAN:DNS-name]' with RSA_EMSA_PKCS1_SHA2_384 successful
```

I have tried all kinds of config options without any success, and I have tried tips from similar treads I have found through different searches. And the exact same config are working on other strongswan-instances running on fedora / debian. 

My gut feeling is that somethings goes wrong with the verification of the certificate. It fails before the EAP-authentication even is tried. But i really cannot understand why it succeeds with the pfsense gateway and not with the mikrotik gateway.  

Any tips, or suggestions, to make sense of this?

**Comment by jfossheim-skyfritt** [comment]
Discussions addressing similar behaviour:

https://wiki.strongswan.org/issues/3343

https://users.strongswan.narkive.com/RDgbVDgo/strongswan-ikev1-pubkey-auth-fails-from-windows-to-linux

https://wiki.strongswan.org/issues/3059

I have also made a dublicate of this post in the mikrotik-forum, hoping that someone there could have some insights:

https://forum.mikrotik.com/viewtopic.php?t=204347

**Comment by tobiasbrunner** [comment, marked as answer]
The MikroTik box seems to not support RFC7427-style signature authentication:

```
feb. 09 12:57:08 [FEDORA-LAPTOP] charon-nm[361273]: 11[IKE] authentication of '[CN=CERTIFICATE COMMON NAME]' with RSA signature successful
```
This means it creates that signature with SHA-1.

That in turn might be disabled on your Rocky Linux box. If SHA-1 is provided via OpenSSL, you could adjust the system-wide policy to allow RSA signatures with SHA-1 in OpenSSL via `sudo update-crypto-policies --set DEFAULT:SHA1` (this basically sets `rh-allow-sha1-signatures = yes` in `openssl.cnf`, so this could theoretically also be set for strongSwan only if you get it to load its own `openssl.cnf` e.g. via `OPENSSL_CONF`).

**Reply by jfossheim-skyfritt** [reply]
Thank you for you're reply: I will test this!

One question: If you take a quick look in the mikrotik forum post: 

(https://forum.mikrotik.com/viewtopic.php?t=204347)

I have tried compiling strongswan both with and without openssl-support: 

> "./configure --disable-gmp --enable-openssl --enable-curl"

And just with: 

> "./configure --enable-curl"

The last one seems to use the gmp-library, would the default-crypto-policy setting be diffrent when not using OpenSSL ? 

Another strange thing, I tried to switch off signature-validation in strongswan as a test, via the options: 

    signature_authentication = no
    signature_authentication_constraints = no

But that dosen't seem to matter? Which sound strange.

**Reply by jfossheim-skyfritt** [reply]
It worked like a charm for the native strongswan-package at least: 

```
Feb 12 11:05:13  [ROCKY-LAPTOP] charon[1181230]: 14[IKE] authentication of '[CN=CERTIFICATE COMMON NAME]' with RSA signature successful
```
Thank you, thank you!!!

**Reply by tobiasbrunner** [reply]
> The last one seems to use the gmp-library, would the default-crypto-policy setting be diffrent when not using OpenSSL ?

Did you remove everything that was installed before you installed the build with default plugins? Because if not, the _openssl_ plugin you installed before will still be there and get loaded.

> Another strange thing, I tried to switch off signature-validation in strongswan as a test, via the options:
> 
> ```
> signature_authentication = no
> signature_authentication_constraints = no
> ```
> But that dosen't seem to matter?

That just disables the signature mechanism the MikroTik doesn't support anyway. So yes, that won't make a difference.

> It worked like a charm for the native strongswan-package at least:

Great!
